### PR TITLE
[Feature]Add Python task "task variable / result transfer" implementation

### PR DESCRIPTION
**Pull request by guyinyou**
Content of PR:
Added VarPool:
Introduction: VarPool is similar to a global variable whose value can be dynamically modified. Variables are passed between different nodes in the task flow.
Difference: Global variables are constant attributes of the task flow, and currently there is no interface that can be dynamically modified.

Precautions:
Need to add the field var_pool in the t_ds_task_instance and t_ds_process_instance tables in the DB, the type is longtext.

How to use:**(Currently only supports Python nodes)**
Value: Just declare the variable name at LocalParams of the node.
Modification: ${setShareVar(variable name, value)} in the script.

Example:

	Node a1:
		Script:
			value = 600 + 60
			${setShareVar(${shareVar1},value+6)}
			
	Node a2:
		Custom parameters:
			shareVar1:"default"
		Script:
			print("shareVar1:",${shareVar1})
		operation result:
			"shareVar1:666"

———————————————————————————————————————————————————
PR的内容：
新增VarPool：
作用：VarPool类似于一个值可动态修改的全局变量，在任务流下不同节点之间传递变量。
区别：全局变量为任务流的常量属性，目前没有接口可以动态修改。

注意事项：
需要在DB中的t_ds_task_instance和t_ds_process_instance表中增加字段var_pool，类型为longtext。

使用：**（目前仅支持Python节点）**
取值：只需在节点的LocalParams处申明变量名。
修改：在脚本中 ${setShareVar(变量名,值)}。

示例：

	节点a1：
		脚本：
			value = 600 + 60
			${setShareVar(${shareVar1},value+6)}

	节点a2：
		自定义参数：
			shareVar1:"default"
		脚本：
			print("shareVar1:",${shareVar1})
		运行结果：
			shareVar1:666

————————————————————————————————————————————————————
Node a1:

————————————————————————————————————————————————————
Node a2:

————————————————————————————————————————————————————
result:

————————————————————————————————————————————————————
Process:
Process

### CODE REVIEW

**Comment by qiaozhanwei** [comment]
please change title to english . thx

**Comment by xingchun-chen** [comment]
e2e failure, Please rebase the dev branch code again

**Comment by simon824** [comment]
I suggest create an issue first and then relation to this pr ,so that you can discuss this feature with others.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/VarPoolUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        ConcurrentHashMap<String, Object> propToValue = new ConcurrentHashMap<String, Object>();
+        VarPoolUtils.convertVarPoolToMap(propToValue, varPool);
+        logger.info(propToValue.toString());
+    }
+    
+    @Test
+    public void testConvertPythonScriptPlaceholders() throws Exception {
+        String rawScript = "print(${p1});\n${setShareVar(${p1},3)};\n${setShareVar(${p2},4)};";
+        rawScript = VarPoolUtils.convertPythonScriptPlaceholders(rawScript);
+        logger.info(rawScript);
```
Comment (simon824) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/VarPoolUtilsTest.java:67]: You need to use Assert to determine whether the result is correct, not just print logs.

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/AbstractCommandExecutor.java [hunk 2]
```diff
@@ -347,8 +351,13 @@ public void run() {
                     long lastFlushTime = System.currentTimeMillis();
 
                     while ((line = inReader.readLine()) != null) {
-                        logBuffer.add(line);
-                        lastFlushTime = flush(lastFlushTime);
+                        if (line.startsWith("${setValue(")) {
+                            varPool.append(line.substring("${setValue(".length(), line.length() - 2));
+                            varPool.append("$guyinyou$");
```
Comment (simon824) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/AbstractCommandExecutor.java:356]: Please replace with another separator here, don't use your name.

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/VarPoolUtils.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     * convertVarPoolToMap
+     * @param propToValue propToValue
+     * @param varPool varPool
+     * @throws ParseException ParseException
+     */
+    public static void convertVarPoolToMap(Map<String, Object> propToValue, String varPool) throws ParseException {
+        if (varPool == null) {
+            return;
+        }
+        String[] splits = varPool.split("\\$guyinyou\\$");
```
Comment (simon824) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/VarPoolUtils.java:83]: Please replace with another separator here, don't use your name.

File: dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/VarPoolUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.server.utils;
+
```
Comment (simon824) [inline dolphinscheduler-server/src/test/java/org/apache/dolphinscheduler/server/utils/VarPoolUtilsTest.java:19]: Keep this test class path consistent with the original class

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/python/PythonTask.java [hunk 5]
```diff
@@ -38,103 +38,110 @@
...
+                        taskExecutionContext.getDefinedParams(),
+                        pythonParameters.getLocalParametersMap(),
+                        CommandType.of(taskExecutionContext.getCmdTypeIfComplement()),
+                        taskExecutionContext.getScheduleTime());
+        
+        try {
+            rawPythonScript = VarPoolUtils.convertPythonScriptPlaceholders(rawPythonScript);
+        }
+        catch (StringIndexOutOfBoundsException e) {
+            // TODO: handle exception
```
Comment (zixi0825) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/worker/task/python/PythonTask.java:129]: If there is no other way to deal with this error, then you should print the error message

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/TaskParams.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        }
+    }
+
+    public void setLocalParamValue(Map<String, Object> propToValue) {
+        if (localParams == null) {
+            return;
+        }
+        for (int i = 0; i < localParams.length; i++) {
+            String prop = localParams[i].get("prop");
+            if (propToValue.containsKey(prop)) {
```
Comment (simon824) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/TaskParams.java:56]: propToValue  may cause Nullpointerexception

**Comment by simon824** [comment]
+1

**Comment by zixi0825** [comment]
+1

**Review by gaojun2048 — CHANGES_REQUESTED** [review]
VarPoolUtils.convertVarPoolToMap(propToValue, processInstance.getVarPool());
            } catch (ParseException e) {
                logger.error("parse {} exception", processInstance.getVarPool(), e);

}
When the processInstance.getVarPool() convert to map exception , we do nothing?
Should this task be marked as failed？

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/TaskParams.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (guyinyou) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/task/TaskParams.java:56]: I have fixed that on commits "7ca9de4"

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java [hunk 7]
```diff
@@ -651,14 +653,22 @@ private void setTaskNodeSkip(List<String> taskNodesSkipList){
...
     private void submitPostNode(String parentNodeName){
 
         List<String> submitTaskNodeList = parsePostNodeList(parentNodeName);
 
         List<TaskInstance> taskInstances = new ArrayList<>();
         for(String taskNode : submitTaskNodeList){
+            try {
+                VarPoolUtils.convertVarPoolToMap(propToValue, processInstance.getVarPool());
+            } catch (ParseException e) {
+                logger.error("parse {} exception", processInstance.getVarPool(), e);
```
[the remark this answers is not in the document]
Comment (guyinyou) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java:666]: Thank you for your proposal. I think about it carefully, but I still can't think of how to deal with the wrong situation. Moreover, under normal circumstances, varpool will only be empty or generated according to the normal format. Generally, there will be no error.

[the remark this answers is not in the document]
Comment (gaojun2048) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java:666]: Some times varpool may updated by other people in some case . So , if the value of varpool not normal and can not convert to map , you can throw a RuntimeException.

[the remark this answers is not in the document]
Comment (guyinyou) [reply dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java:666]: You're right. I'll take it

**Review by gaojun2048 — APPROVED** [review]

**Review by gaojun2048 — APPROVED** [review]
+1

File: dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java [hunk 8]
```diff
@@ -651,14 +653,23 @@ private void setTaskNodeSkip(List<String> taskNodesSkipList){
...
+                throw new RuntimeException();
+            }
+            TaskNode taskNodeObject = dag.getNode(taskNode);
+            VarPoolUtils.setTaskNodeLocalParams(taskNodeObject, propToValue);
```
Comment (hei12138) [inline dolphinscheduler-server/src/main/java/org/apache/dolphinscheduler/server/master/runner/MasterExecThread.java:670]: there is bug will cause the origin param be replace by the task param,so the task will get nothing when use  taskExecutionContext.getTaskParams()

**Comment by luojieio** [comment]
why I use 1.3.8 release version can not do it ?

**Comment by luojieio** [comment]
image

**Pull request merged**